cli/azd/pkg/commands/builder.go

```diff
@@ -12,15 +12,37 @@ import (
 	"github.com/spf13/cobra"
 )
 
+type BuildOptions struct {
```

Comment:
I wonder if we would want a constructor for this ???  Or some mechanism to set arguments as mandatory.

This is because moving from the current approach to using the options-bag makes all these values `optional`.  I can pass an empty `BuildOptions{}` to `Build()` and what does it mean?  

If we are going this path, we need a `pre-condition` to panic if these values are not provided to `Build()` 

IMO, option-bags should be used only for `optional` arguments which can be defaulted to something.  (That's the guideline for SDK clients)

Reply:
> I wonder if we would want a constructor for this ??? Or some mechanism to set arguments as mandatory.

I'm not entirely thrilled about the constructor approach -- we seem to create a significant amount of indirection to accomplish something simple.

> If we are going this path, we need a `pre-condition` to panic if these values are not provided to `Build()`

This might be independent of whether a bag is used. For example, we did not perform `nil` check on `rootOptions`. Such is possible in old and new code paths.

> IMO, option-bags should be used only for `optional` arguments which can be defaulted to something. (That's the guideline for SDK clients)

I definitely see why it's important to be extremely clear and consistent in the design language of the SDK clients. We *could* be less strict here as this isn't a public API. That being said, we can definitely adhere to a higher standard to avoid potential developer errors.

If we do want to go down this stricter path, I think moving all optional arguments to the struct makes sense and keeping all required arguments as part of the method args. 

There is an argument to be made: If some parameter is required and part of the contract, it **should** be hard to make a change and keeping it as part of the method args is inline with that.

Reply:
Agree.
A constructor is always ugly. LoL.

I created a PR in the past to update the `build()` function to a complete builder pattern.  
The PR is part of the old repo (https://github.com/Azure/azure-dev-pr/pull/1176)

That's yet another alternative :)

At any case, I like this:

 >  moving all optional arguments to the struct makes sense and keeping all required arguments as part of the method args.

Let's wait on @wbreza and @ellismg

Reply:
I definitely agree with the `Builder` pattern being a good idea if we start adding a lot of complicated optionals in the future. For a short-term fix, I think having optionals broken up is a good step in the right direction.

We can easily pull out the options into `Set<Option>` methods on the builder as a next step. Pulling out optionals sets up nicely to transform into a Builder.